File: frontend/src/types/sale.ts
export enum SaleStatus {
  PENDING = 'pending',
  COMPLETED = 'completed',
  CANCELLED = 'cancelled',
  RETURNED = 'returned',
}

export interface SaleItem {
  id?: string;
  sale_id?: string;
  product_id: string;
  product_name?: string;
  product_code?: string;
  brand_name?: string;
  product_image_url?: string;
  size?: string;
  quantity: number;
  seller_sale_price_original: number; // 판매자 판매가격(원본 통화)
  seller_sale_currency?: string; // 판매 국가/통화
  seller_sale_price_krw: number; // 판매자 판매가격(한국 환율)
  company_sale_price?: number; // 회사 판매가격 (관리자 입력)
  seller_margin?: number; // 판매자 마진 (자동 계산)
  created_at?: string;
  updated_at?: string;
  product?: any; // Product 타입 참조
}

export interface SaleItemCreate {
  product_id: string;
  product_name?: string;
  size?: string;
  quantity: number;
  seller_sale_price_original: number;
  seller_sale_currency?: string;
  seller_sale_price_krw: number;
  product_image_url?: string;
}

export interface Sale {
  id?: string;
  sale_number?: string; // 판매번호 (S로 시작)
  sale_date: string;
  seller_id?: string;
  seller_name?: string; // 판매자명 표시용
  customer_name?: string; // 고객명 또는 고객처
  customer_contact?: string;
  total_seller_amount?: number; // 총 판매자 판매금액
  total_company_amount?: number; // 총 회사 판매금액
  total_seller_margin?: number; // 총 판매자 마진
  status?: SaleStatus;
  notes?: string;
  transaction_statement_url?: string; // 거래명세서 URL
  tax_invoice_url?: string; // 세금계산서 URL
  created_at?: string;
  updated_at?: string;
  items?: SaleItem[];
}

export interface SaleCreate {
  sale_date: string;
  customer_name?: string;
  customer_contact?: string;
  notes?: string;
  transaction_statement_url?: string;
  tax_invoice_url?: string;
  items: SaleItemCreate[];
}

export interface SaleUpdate {
  sale_date?: string;
  customer_name?: string;
  customer_contact?: string;
  status?: SaleStatus;
  notes?: string;
  transaction_statement_url?: string;
  tax_invoice_url?: string;
  total_seller_amount?: number;
  total_company_amount?: number;
  total_seller_margin?: number;
}

export interface SaleList {
  total: number;
  items: Sale[];
}

export interface SaleListParams {
  skip?: number;
  limit?: number;
  start_date?: string;
  end_date?: string;
  status?: SaleStatus | SaleStatus[];
  brand_name?: string | string[];
  search?: string; // 고객명, 연락처 검색
}

File: backend/app/schemas/sale.py
from typing import Optional, List
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID
from pydantic import BaseModel, field_validator
from app.models.sale import SaleStatus

class SaleItemBase(BaseModel):
    product_id: str
    product_name: Optional[str] = None
    product_code: Optional[str] = None
    brand_name: Optional[str] = None
    category: Optional[str] = None
    size: Optional[str] = None
    quantity: int = 1
    seller_sale_price_original: Decimal
    seller_sale_currency: Optional[str] = "USD"
    seller_sale_price_krw: Decimal
    product_image_url: Optional[str] = None

class SaleItemCreate(SaleItemBase):
    pass

class SaleItemUpdate(SaleItemBase):
    product_id: Optional[str] = None
    quantity: Optional[int] = None
    seller_sale_price_krw: Optional[Decimal] = None

class SaleItem(SaleItemBase):
    id: str
    sale_id: str
    company_sale_price: Optional[Decimal] = None
    seller_margin: Optional[Decimal] = None
    created_at: datetime
    updated_at: datetime

    @field_validator('id', 'sale_id', 'product_id', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v):
        if isinstance(v, UUID):
            return str(v)
        return v

    class Config:
        from_attributes = True

class SaleBase(BaseModel):
    sale_date: date
    customer_name: Optional[str] = None
    customer_contact: Optional[str] = None
    notes: Optional[str] = None
    transaction_statement_url: Optional[str] = None
    tax_invoice_url: Optional[str] = None

class SaleCreate(SaleBase):
    items: List[SaleItemCreate]

class SaleUpdate(BaseModel):
    sale_date: Optional[date] = None
    customer_name: Optional[str] = None
    customer_contact: Optional[str] = None
    total_company_amount: Optional[Decimal] = None
    total_seller_margin: Optional[Decimal] = None
    status: Optional[SaleStatus] = None
    notes: Optional[str] = None
    transaction_statement_url: Optional[str] = None
    tax_invoice_url: Optional[str] = None

class Sale(SaleBase):
    id: str
    sale_number: str
    seller_id: str
    seller_name: Optional[str] = None
    total_seller_amount: Optional[Decimal] = None
    total_company_amount: Optional[Decimal] = None
    total_seller_margin: Optional[Decimal] = None
    status: SaleStatus
    created_at: datetime
    updated_at: datetime
    items: List[SaleItem] = []

    @field_validator('id', 'seller_id', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v):
        if isinstance(v, UUID):
            return str(v)
        return v

    class Config:
        from_attributes = True

class SaleList(BaseModel):
    total: int
    items: List[Sale]